### PR TITLE
fix(errors): Add error handling for when we hit the rate limit

### PR DESCRIPTION
**Pull request by craigtweedy**
## Status

* ✅ Ready

## Type

* ### Fix
  Fixes a bug

## Description

The SDK does not correctly handle issues where JSON is not returned in the body of the request, and as such bails out when the rate limit is hit. This pull request implements an additional check for a 429 status code and fills in the JSON appropriately.

## Notes

Handling individually error codes like this may get tricky if there are multiple error codes which don't return a JSON body - which could exist if there were to be an issue with the API, or the API implements such a code which doesn't return any JSON (like it has done here). In the future, this should likely expand out into more robust error handling which accounts for all cases, as well as return additional information this may not be present in the body, such as status codes or headers so that we don't have to fill in the JSON body.

### CODE REVIEW

File: src/utils/helpers.js [hunk 1]
```diff
@@ -58,6 +58,22 @@ export function parseJSON(response) {
     })
   }
 
+  if (response.status === 429) {
+    // Only the body of the response is passed back so to catch a 429
+    // we fill the JSON body with some error details to give information
+    // to the consumer
+    const error = {
+      errors: [{ status: 429 }]
```
Comment (ynnoj) [inline src/utils/helpers.js]: @craigtweedy Not want to return a `title` or `detail`/`message` here? `429` is quite ambiguous.

File: src/utils/helpers.js [hunk 2]
```diff
@@ -58,6 +58,22 @@ export function parseJSON(response) {
...
+    }
+    return new Promise(resolve => {
```
Comment (ynnoj) [inline src/utils/helpers.js]: @craigtweedy We can tidy this whole function up really - no need for 3 separate returns based on `status` logic.

File: src/utils/helpers.js [hunk 3]
```diff
@@ -58,6 +58,22 @@ export function parseJSON(response) {
...
+      resolve({
+        status: response.status,
```
Comment (ynnoj) [inline src/utils/helpers.js]: @craigtweedy Destructure? 😝

File: src/utils/helpers.js (hunk 1 above)
Comment (craigtweedy) [reply src/utils/helpers.js]: Why do you feel 429 is ambiguous? Surely it only means the one thing?

I thought about a title and detail here - but didn't want to start the task of properly defining our own error codes and messages.

Comment (craigtweedy) [reply src/utils/helpers.js]: Then again, the status may not always be the same as the HTTP status code, and to be clearer for devs that may not know that's a rate limit it is probably best to be clear.

File: src/utils/helpers.js (hunk 2 above)
Comment (craigtweedy) [reply src/utils/helpers.js]: I agree - was mostly to fit into the style already there but I will update this entire function.

File: src/utils/helpers.js (hunk 3 above)
Comment (craigtweedy) [reply src/utils/helpers.js]: How do you mean in this instance? Could you add a code suggestion?

File: src/utils/helpers.js (hunk 2 above)
Comment (craigtweedy) [reply src/utils/helpers.js]: How about this? 

```js
export function parseJSON(response) {

  return new Promise(resolve => {
    response.text().then(body => {

      let json = '{}'

      if (body !== '') {
        json = JSON.parse(body)
      } 
      
      if (!response.ok) {
        json = { errors: [{ status: response.status }] }
      }

      resolve({
        status: response.status,
        ok: response.ok,
        json
      })
    })
  })
}
```

Comment (ynnoj) [reply src/utils/helpers.js]: @craigtweedy Looks better, but not sure why `text()` is needed in this instance? Maybe I'm misunderstanding.

```js
export function parseJSON(response) {
  return new Promise(resolve => {
    const { status, ok } = response

    response.json().then(payload => 
      let json = payload

      if (!ok) {
        json = { errors: [{ status }] }
      }

      if (status === 204) {
        json = '{}'
      }

      resolve({
        status,
        ok,
        json
      })
    )
  }) 
}
```

File: src/utils/helpers.js (hunk 3 above)
Comment (ynnoj) [reply src/utils/helpers.js]: > How do you mean in this instance? Could you add a code suggestion?

@craigtweedy I was mostly trolling, but see above example.

File: src/utils/helpers.js (hunk 2 above)
Comment (craigtweedy) [reply src/utils/helpers.js]: `response.json()` throws an error if the body is empty - which it is in a couple of situations like when we hit the rate limit. Your example above would still cause the SDK to bail out when the rate limit is hit. 

For this reason, I figured parsing as text and then filling in the empty body would be the path of least resistance.

**Comment by craigtweedy** [comment]
The above comments about the error function need to be added in, so I'll do that in today and update the branch from master, then you can merge when you like.

**Pull request merged**